### PR TITLE
Updates to container manager and internal container lifecycle to accommodate TopologyManager

### PR DESCRIPTION
**Pull request by lmdaly**
**What this PR does / why we need it**:
Adding support for Topology Manager to Container Manager and Internal container lifecycle
Topology Manager based merged design proposal here: kubernetes/community#1680
Issue for tracking PRs: #72828
**What type of PR is this?**

> /kind feature

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Comment by spiffxp** [comment]
/hold
ref: https://github.com/kubernetes/enhancements/issues/693#issuecomment-466728227

/ok-to-test

**Review by klueska — APPROVED** [review]
LGTM pending previous PRs are merged.

**Comment by lmdaly** [comment]
PR Depends on:
https://github.com/kubernetes/kubernetes/pull/73580
https://github.com/kubernetes/kubernetes/pull/74411

File: pkg/kubelet/cm/container_manager_linux.go
Comment (sjenning) [inline pkg/kubelet/cm/container_manager_linux.go]: In the other OSes, we return a nil `topologyManager`.  I don't think we want to use a fake manager outside of test code right?

File: pkg/kubelet/cm/container_manager_linux.go
Comment (sjenning) [inline pkg/kubelet/cm/container_manager_linux.go]: this message could conflict if the policy name in the flag is invalid and `NewManager` falls back to the default policy

File: pkg/kubelet/cm/container_manager_linux.go
Comment (sjenning) [inline pkg/kubelet/cm/container_manager_linux.go]: this seems like debug code

File: pkg/kubelet/cm/container_manager_linux.go
Comment (klueska) [reply pkg/kubelet/cm/container_manager_linux.go]: +1

File: pkg/kubelet/cm/container_manager_linux.go
Comment (klueska) [reply pkg/kubelet/cm/container_manager_linux.go]: I would argue that falling back to the default policy on an invalid policy name in the flag is bad behaviour. It should instead return an error, and fail fast in this case.

Comment (sjenning) [reply pkg/kubelet/cm/container_manager_linux.go]: i agree, this should be fatal to the kubelet, `NewManager` should return a `(Manager, err)`

Comment (klueska) [reply pkg/kubelet/cm/container_manager_linux.go]: I think this was borrowed from the semantics of the CPUManager, which currently does a similar fallback to the None policy if an incorrect name is passed in the flag.

It's probably not worth blocking this PR until this is in place, since (at the moment) it's just potential logging error.

I will open a followup PR once this is merged to address this (in both the TopologyManager and the CPUManager).

File: pkg/kubelet/cm/container_manager_linux.go
Comment (lmdaly) [reply pkg/kubelet/cm/container_manager_linux.go]: From what I can see, the other OSes don't have TopologyManager is part of the containerManagerImpl so there is no reference, so we don't return anything?

 I was following the convention used by DeviceManager which returns a ManagerStub below. There is also a reference to the Fake CPU Manager in the other OSes when returning the internalContainerLifecycleImpl.

If I return nil, the code errors as cm.TopologyManager functions are called below. So I could wrap those calls in a check for the feature gate but looks like the fake/stub managers are used outside of test code.

File: pkg/kubelet/cm/container_manager_linux.go
Comment (lmdaly) [reply pkg/kubelet/cm/container_manager_linux.go]: Good spot, it is. Will remove, thanks!

File: pkg/kubelet/cm/container_manager_linux.go
Comment (lmdaly) [reply pkg/kubelet/cm/container_manager_linux.go]: Going to resolve this conversation as there is a follow on PR planned. Thanks!

Comment (klueska) [reply pkg/kubelet/cm/container_manager_linux.go]: Here is the PR: https://github.com/kubernetes/kubernetes/pull/80294

Depending on which one lands first (either this PR or that one), one of them will need to be updated to make the change that handles `NewManager()` returning an `error` at this point in the code.

File: pkg/kubelet/cm/container_manager_linux.go
Comment (klueska) [reply pkg/kubelet/cm/container_manager_linux.go]: I actually think the only call you would need to wrap is this one (which probably makes sense to wrap anyway):

```
pkg/kubelet/kubelet.go:
klet.admitHandlers.AddPodAdmitHandler(klet.containerManager.GetTopologyPodAdmitHandler())
```

Looking through the code, I don't see anywhere else that `cm.topologyManager` is referenced (other than https://github.com/kubernetes/kubernetes/pull/74357/files#diff-06aa1b61a9bc4a50aa8d8b0a69d95bb2R347), whose internal references to the `topologyManager` are already protected by a check on the feature gate:

https://github.com/kubernetes/kubernetes/pull/74357/files#diff-81cef540f416b0451c18470420adc42dR40

Comment (lmdaly) [reply pkg/kubelet/cm/container_manager_linux.go]: Ah yes I see the code I was referring to is actually in the CPU and Device Manager PRs.
https://github.com/kubernetes/kubernetes/pull/73920 & https://github.com/kubernetes/kubernetes/pull/74345 have the references to topology manager to register them as Hint Providers

Comment (klueska) [reply pkg/kubelet/cm/container_manager_linux.go]: It still feels more natural to me to remove the `FakeManager()` here and protect all of the accesses to `cm.topologymanager` with a check on the feature gate, but I will leave it to @sjennings for the final call.

Comment (sjenning) [reply pkg/kubelet/cm/container_manager_linux.go]: we can clean it up in a follow on.  there are a number of things that can be cleaned up.  i'm making a list :)

**Comment by sjenning** [comment]
ah yes, we are out of `pkg/kubelet/cm` now. my powers are useless here!
/assign @dchen1107 @derekwaynecarr

**Comment by klueska** [comment]
Looks like it needs a rebase on the newest master before retesting

**Comment by derekwaynecarr** [comment]
taking a look now per sig-node discussion.
/assign

File: pkg/kubelet/cm/container_manager.go
Comment (derekwaynecarr) [inline pkg/kubelet/cm/container_manager.go]: I suspect we will plan to move this out of experimental in the future, is it better if we just call this `TopologyManagerPolicy` to avoid a future rename?

File: pkg/kubelet/cm/container_manager_linux.go
Comment (derekwaynecarr) [inline pkg/kubelet/cm/container_manager_linux.go]: nit: "Initializing"

**Review by derekwaynecarr — APPROVED** [review]
/approve
/lgtm

in a follow-on please remove the use of `Experimental` prefix, it ends up being a pain when they no longer are experimental to clean the naming up later.

File: pkg/kubelet/cm/container_manager.go
Comment (tedyu) [reply pkg/kubelet/cm/container_manager.go]: +1

**Pull request merged**
